ForceMove safety (#309)
statechannels/monorepo: packages/nitro-protocol/docs/force-move-safety.md

> | Max       | Width                | Table          | Width        |

Author: I'll temporarily move it out of `docs` once the comments are addressed.

This is the maxiumum width of a nicely formatted table in a medium post.

# Intro

ForceMove is a state-channel protocol, designed to be a simple as possible.
A set of participants exchanging states in private, which determine some outcome of the state channel.
An adjudicator, in the form of a smart contract, enforces the outcome on a blockchain, which would typically involve some financial transaction.
As such, the participants wish for the system to protect against malicious actors that try to prevent the most fair outcome from being recorded on-chain.

At Magmo, we have spent some time specifying the ForceMove protocol in [ TLA+ ](https://en.wikipedia.org/wiki/TLA%2B), a formal specification language based on set theory.
Engineers at Amazon Web Services [have used](https://lamport.azurewebsites.net/tla/formal-methods-amazon.pdf) TLA+ since 2011 when designing some of their critical systems\*.

# ForceMove overview

<!-- TODO: I'm not sure what to say in the general overview -->
<!-- Currently:
1. A ForceMove state channel is a state machine that determines the distribution of assets
2. Valid transitions, turn numbers, ordered turns
3. Adjudicator's role
4. Supported states
-->

> In a ForceMove state channel, participants, privately update the state of some ledger.

Reviewer: IMO 'ledger' seems a bit outdated these days and I don't see it used an awful lot in the ethereum community.

```suggestion
In a ForceMove state channel, a small number of participants update a private state that is shared between them.
```

> The state's outcome dictates how certain assets are allocated.

Reviewer: ```suggestion
The state's outcome ultimately dictates how the (public) blockchain state is updated: for example, how certain assets are allocated.
```

> The ledger may have additional state relative to the type of application the state channel is meant to support.

Reviewer: ```suggestion
The state may include additional data relating to the type of application the state channel is meant to support.
```

In this blog post, we restrict to state channels with two participants, Alice and Bob.
They update the ledger by exchanging signed copies of what they agree on as the _current state_ of the state channel.
They only sign updates to the ledger according to a pre-determined set of rules: they agree in advance to a function `validTransition(s, s')`, and a transition from `s` to `s'` is only allowed when `validTransition(s, s')` returns `true`.
For example, in a state channel used to play a game of chess, `validTransition` would only return true if the transition represents a legal chess move.
Each state includes a `turnNumber`, which must increase by one on each valid transition.
This allows for newer states to take precedence over older states.

Participants are ordered, and turns are taken in order.
In our case, for example, Alice decides what state to transition to on even turns, and Bob decides on odd turns.
We say that a state `s'` is _supported_ if there is a pair of states `(s, s')` such that

- `validTransition(s, s') == true`
- `s` and `s'` are both signed by the correct participant

> In this case, `(s, s')`is called _support_ for the state`s'`.

Reviewer: ```suggestion
In this case, `(s, s')`is called _support_ for the state `s'`.
```

> This means that both participants have indicated their willingness to conclude the channel and distribute the assets according to`s'.outcome`.

Reviewer: ```suggestion
This means that both participants have indicated their willingness to conclude the channel and distribute the assets according to `s'.outcome`.
```

> A smart contract called the _adjudicator_ is used to manage the assets, fairly dividing the assets at the conclusion of the state channel.

Reviewer: Also, in the latest revision of the contracts, asset holding has been split off into contracts which are separate from the adjudicator. You may justifiably want to gloss over that here, but it is a potential source of confusion. You might replace "manage the assets" with "execute some change in the blockchain state" which is still true under the current contracts arrangement. (The asset holders merely respond to that change).

The purpose of `forceMove` action is to move to a `Challenge` state.
The three remaining actions clear the challenge, moving to an `Open` state, by either

- providing a later, supported state (`respond` & `respondWithAlternativeMove`), and thus progressing the channel; or
- prove that the challenger acted in bad faith (`refute`)

In addition, the adjudicator has one further action, `timeout(S)`, where `S` is is some on-chain channel state.
The purpose of `timeout` is to close the channel by moving to the `Closed` state.
In practice, this action is implicit, and occurs with the passage of time.

## `forceMove`

To apply this action, a participant must supply a supported state `s` with turnNumber `m`.
The participant must also authorize this action by signing a special message.
Therefore, the action also includes the challenger `p`.

## `respond`

To apply this action, a participant can simply supply a correctly-signed state `s`.
This state should form a valid transition from the current `challengeState`.

## `respondAlt`

This action allows participants to provide the adjudicator an alternative set of states that also increases the `turnNumRecord`.

> Participants may wish to sign multiple states with the same `turnNumRecord` for various practical purposes.

Reviewer: ```suggestion
Participants may wish to sign multiple states with the same `turnNumber` for various practical purposes.
```

If the challenger signed such a state, the adjudicator can infer that the challenger should be aware of a later supported state, and discards the current challenge.
At this point, the adjudicator could penalize the challenger for malicious intent.

## `timeout`

Whenever the state of the channel is updated to `S`, the adjudicator makes a note of `S`, and starts a timer.
When the timer runs out, the adjudicator applies `timeout(S)`.

In practice, this transition happens implicitly: when in a `Challenge` state, the challenge expiration time is recorded in a field `finalizesAt`, and once `finalizesAt <= now`, the channel is closed.

## State Transitions, V1

We record the allowed state transitions from the original ForceMove protocol in the following table.
By default, any (state, action) pair not listed below does not change the state.

| State       | Action           | Next state  | Requiremeents   |
| ----------- | ---------------- | ----------- | --------------- |
| Open(n)     | forceMove(m,s,p) | Chal(m,s,p) | m >= n          |
| Chal(n,s,p) | respond(s')      | Open(n+1)   | s -> s'         |

> | Chal(n,s,p) | respondAlt(n+1)  | Open(n+1)   |                 |

Reviewer: The argument of this function is of a different type to that specified in the function definition above on L75 (state vs. turnNum).

| Chal(n,s,p) | refute(m,s')     | Open(n)     | m > n           |
|             |                  |             | signer(p') == p |
| Chal(n,s,p) | timeout(S)       | Closed(s)   | matches(S)      |

Here, `matches(S)` indicates that the current on-chain channel state is `S`.
`Chal` is used as a short-hand for `Challenge`.
In other words, applying `timeout` to a `Challenge` state only has an effect if the state has not changed since the adjudicator timer has started.

> Applying `timeout` to an open state has no effect, as the transition is not listed in this table.

Reviewer: It would be great to disambiguate the state of the _channel_ from the state of the _adjudicator_ throughout. For example, "open adjudicator state" and so on.

Author: I've done this partway. I'm 50/50 on if "state" needs any further clarification in the phrase "`Open` state". I'm almost thinking about introducing some variable called `AdjudicatorState` to be completely explicit, since there are so many different types of states discussed in the post.

Reviewer: Yeah it's a judgement call on whether the context is enough to prevent confusion, or whether more terminology helps. I tend to prefer the more explicit route, because it is easy to underestimate  the confusion for those who don't look at this every day.

We use the notation `A >> S` to denote the state resulting in applying action `A` to state `S`.

# Safety

Since the channel only closes from a `Challenge` state, and the channel only enters a challenge state when the adjudicator sees a supported state, Alice must have been, at some point in time, satisfied with any state that the channel closes with.

> However, it may not close with the latest state.

Reviewer: ```suggestion
However, there may be a later state which should supersede the challenge state; if Alice (or some other party) does not act, the channel may close in the out-of-date challenge state.
```

Can Alice guarantee that, in a constant number of actions, she can ensure the channel can only close with a state at turn at least `n`?
She can trivially accomplish this if she can move the state's `turnNumRecord` to be at least `n`, since the `turnNumRecord` is non-decreasing:

```
forceMove(n,s,Alice) >> Open(m) == Challenge(n,s,Alice)
```

However, in the context of the Ethereum blockchain, this is not sufficient.
Alice applies `forceMove` by calling a state-modifying function on an Ethereum smart contract in some transaction.
The transaction is not recorded immediately, and once Bob sees her transaction, he might attempt to front-run her transaction with a different action in his own transaction.
If his transaction is recorded first, it might change the state in such a way that Alice's transaction reverts.

# Gas-optimization

The situation is worsened by an optimization we've done to save on gas fees.
A hash of the entire channel state is stored in a single `bytes32` slot on-chain.
Therefore, when applying an action, the participant must provide the current state as a part of the calldata.

> If the hash of the provide state does not match the stored value, the action has no effect.

Reviewer: ```suggestion
If the hash of the provided state does not match the stored value, the action has no effect.
```

║ Chal(n,s,p) ║ refute(m,s',S)     ║ Open(n)     ║ matches(S)   ║
║             ║                    ║             ╠══════════════╣
║             ║                    ║             ║ m > n        ║
║             ║                    ║             ╠══════════════╣
║             ║                    ║             ║ p == sig(s') ║
╠═════════════╬════════════════════╬═════════════╬══════════════╣
║ Chal(n,s,p) ║ respond(s',S)      ║ Open(n+1)   ║ matches(S)   ║
║             ║                    ║             ╠══════════════╣
║             ║                    ║             ║ s->s'        ║
╠═════════════╬════════════════════╬═════════════╬══════════════╣
║ Open(n)     ║ respondAlt(S)      ║ Open(n+1)   ║ matches(S)   ║
╚═════════════╩════════════════════╩═════════════╩══════════════╝
```

```
function matches(S) {
  return keccack256(S) == currentState;
}
```

## Attack

- Suppose the channel is in `Open(k)`
- Alice wishes to apply `forceMove(n,s,Alice)`
- Bob front-runs with `forceMove(k,s',Eve)`

```
   forceMove(n,s,Alice,S) >> forceMove(k,s',Eve,S) >> Open(k)
== forceMove(n,s,Alice,S) >> Challenge(k,s',Eve)
== Challenge(k,s',Eve)
```

> 

Reviewer: I think this notation is great; but it might be nice to remind the reader here and explicitly mention that `A>>B>>S` denotes tx `A` being front-run by tx `B`.

Reviewer 2: agree - that took me a while to work out

The problem is twofold:

- The transition `forceMove >> Challenge` is not allowed in the current version of the protocol

> - Eve if it were, `S` no longer matches the current state, and thus, no action would change the channel state

Reviewer: ```suggestion
- Even if it were, `S` no longer matches the current state, and thus, no action would change the channel state
```

  - Indeed, we were already aware of one problem exposed by TLA+
  - Yet, TLA+ serves well as a way to force us to explicitly specify the protocol in a way that's amenable to checking for violating behaviours -->

# Version 3

While trying to model the behaviour of `respondWithAlternativeMove` in TLA+, we realized that this action had an artificial limitation.
There's no reason why it must advance the `turnNumRecord` by exactly one.
This restriction complicates state management, since the user may need to store stale states in order to use them to `respondWithAlternativeMove`.
Moreover, there's no reason why it can't be applied to an open state to increase the `turnNumRecord`.
Thus, we replaced it with the `checkpoint` action.
This allows one to only care about the latest supported state, and it can be used to protect one's assets before going offline for a period of time.

| State       | Action             | NextState   | Requirements |
| ----------- | ------------------ | ----------- | ------------ |
| Open(n)     | forceMove(m,s,p,S) | Chal(m,s,p) | matches(S)   |
|             |                    |             | m >= n       |

> | Open(n,S)   | checkpoint(m,S)    | Open(m))    | matches(S)   |

Reviewer: ```suggestion
| Open(n,S)   | checkpoint(m,S)    | Open(m)     | matches(S)   |
```

|             |                    |             | m > n        |
| Chal(n,s,p) | checkpoint(m,S)    | Open(m)     | matches(S)   |
|             |                    |             | m > n        |
| Chal(n,s,p) | refute(m,s',S)     | Open(n)     | matches(S)   |
|             |                    |             | m > n        |
|             |                    |             | p == sig(s') |
| Chal(n,s,p) | respond(s',S)      | Open(n+1)   | matches(S)   |
|             |                    |             | s->s'        |

> This offers Alice a new strategy, applying `checkpoint(n, t, S)`, regardless of the current state.

Reviewer: ```suggestion
This offers Alice a new strategy, applying `checkpoint(n, t, S)`, regardless of the current adjudicator state.
```

> - Alice can call `checkpoint(n,t,S)` at any time with her most recent state,which has turn `n`

Reviewer: does "at any time" include when the channel is in challenge mode?

> - If this transaction gets recorded,it moves to `Open(n)`

Reviewer: recorded -> mined?

- The only valid actions from this point are
  - `checkpoint(m,S') >> Open(n) == Open(m)` for some `m > n`
  - `forceMove(m,s,p,S') >> Open(n) == Challenge(m,s,p)` for some `m > n`

> - Alice is happy

Reviewer: ...because

| ----------- | ---------------- | ----------- | ------------ |
| Open(n)     | forceMove(m,s,p) | Chal(m,s,p) | m >= n       |
| Open(n,S)   | checkpoint(m)    | Open(m)     | m > n        |
| Chal(n,s,p) | checkpoint(m)    | Open(m)     | m > n        |
| Chal(n,s,p) | refute(m,s',S)   | Open(n)     | matches(S)   |
|             |                  |             | m > n        |
|             |                  |             | p == sig(s') |
| Chal(n,s,p) | respond(s',S)    | Open(n+1)   | matches(S)   |
|             |                  |             | s->s'        |

In this version of the protocol, the `matches` function is implemented in Solidity as follows.

```
function matches(ChannelState memory S) public pure returns (bool) {
  // Casting from uint256 to uint160 returns the 160 least significant bits.
  // The 96 most significant bits can be used to store un-hashed data, including
  // the turnNumRecord
  return uint160(uint256(keccack256(S))) == uint160(uint256(currentState)) ;
}
```

Thus, the `forceMove` and `checkpoint` actions are always enabled, if the participant can provide a newer state than the current `turnNumRecord`.

> This counters the attacks found on the previous versions, thanks to the commutativity of the `forceMove` and `checkpoint` actions:

Reviewer: Perhaps include a subclause to remind the uninitiated about what "commutativity" means.

- Suppose we're in `Open(0)`
- Alice wants to apply `checkpoint(n)`
- Bob front-run with `forceMove(0,s,p)`

```
   checkpoint(n) >> forceMove(0,s,Eve) >> Open(0)
== checkpoint(n) >> Challenge(0,s,Eve)
== Open(n)
```

However, TLA+ exposed a new attack, unknown to us at the time.

## Attack

- Suppose we're in `Open(n)`, and alice's latest state is `s`, with turn `n`.
- Alice wants to apply `forceMove(n,s,Alice)`
- Bob front-runs with `forceMove(n,s,Eve)`

```
   forceMove(n,s,Alice) >> forceMove(n,s,Eve) >> Open(n)
== forceMove(n,s,Alice) >> Challenge(n,s,Eve)
== Challenge(n,s,Eve)
```

- Alice has no signed state later than `n`, so she cannot perform any action.
- Bob can sign any state `s'` with her turn number, and she calls `refute(m,s',S)`.

```
refute(m,s',S) >> Challenge(n,s,Eve) == Open(n)
```

> - The state has returns to `Open(n)`.

Reviewer: ```suggestion
- The state has returned to `Open(n)`.
```

While there are techniques that can block this attack, such as storing who has challenged so far with the current `turnNumRecord`, we decided to forgo the `refute` action altogether until we have a clear application for it.

Note that this attack was present in all of the above versions.
Prior to writing specs in TLA+, we were more or less aware of the previous attacks, but we had not thought about this simple attack.

# Version 5

This leaves us with the current version of the protocol:

| State       | Action           | NextState   | Requirements |
| ----------- | ---------------- | ----------- | ------------ |
| Open(n)     | forceMove(m,s,p) | Chal(m,s,p) | m >= n       |
| Open(n,S)   | checkpoint(m)    | Open(m))    | m > n        |
| Chal(n,s,p) | checkpoint(m)    | Open(m)     | m > n        |
| Chal(n,s,p) | respond(s',S)    | Open(n+1)   | matches(S)   |
|             |                  |             | s->s'        |

With this final protocol, when Alice has a state `s` of turn `n`, she has two strategies to protect her assets.

If she wishes to force Bob to move, she applies `forceMove(n, s, Alice)`.

> When her transaction is proccessed, we can partition the possible states into four cases:

Reviewer: ```suggestion
When her transaction is proccessed, we can partition the possible adjudicator states into four cases:
```

1. `Open(m)`
   a. `m <= n`
   b. `m > n`
2. `Challenge(m, s', p')`
   a. `m < n`
   b. `m >= n`

In cases 1a & 2a, her transaction succeeds, and the state changes to `Challenge(n, s, Alice)`.
In cases 1b & 2b, her transaction fails, but she has seen a supported state `s'` that trumps `s`, which was her goal.

If she wishes to protect her assets before going offline, she applies `checkpoint(n)`.

> When her transaction is proccessed, we can partition the possible states into four cases:

Reviewer: ```suggestion
When her transaction is proccessed, once again we can partition the possible states into four cases:
```

| State       | Action             | NextState   | Requirements |
| ----------- | ------------------ | ----------- | ------------ |
| Open(n)     | forceMove(m,s,p,S) | Chal(m,s,p) | matches(S)   |
|             |                    |             | m >= n       |
| Open(n,S)   | checkpoint(m,S)    | Open(m))    | matches(S)   |
|             |                    |             | m > n        |
| Chal(n,s,p) | checkpoint(m,S)    | Open(m)     | matches(S)   |
|             |                    |             | m > n        |
| Chal(n,s,p) | refute(m,s',S)     | Open(n)     | matches(S)   |
|             |                    |             | m > n        |
|             |                    |             | p == sig(s') |
| Chal(n,s,p) | respond(s',S)      | Open(n+1)   | matches(S)   |
|             |                    |             | s->s'        |

This offers Alice a new strategy, applying `checkpoint(n, t, S)`, regardless of the current state.

- Alice can call `checkpoint(n,t,S)` at any time with her most recent state,which has turn `n`

> - If this transaction gets recorded,it moves to `Open(n)`

Author: ```suggestion
- If this transaction gets mined, it moves to `Open(n)`
```
